derive-c-marshalling/src/lib.rs

```diff
@@ -0,0 +1,167 @@
+                        &cstring_type
+                    } else {
+                        &field.ty
+                    };
+                    quote! {
+                        #ident: #ty
+                    }
+                });
+            let field_pointer_declarations = fields
+                .iter()
+                .map(|field| {
+                    let ident = &field.ident.as_ref().unwrap();
+                    let ty = if field.ty == absolute_string_type || field.ty == relative_string_type {
+                        &cptr_type
+                    } else {
+                        &field.ty
+                    };
+                    quote! {
+                        #ident: #ty
+                    }
+                });
+            let c_field_initializers = fields
+                .iter()
+                .map(|field| {
+                    let ident = &field.ident.as_ref().unwrap();
+                    if field.ty == absolute_string_type || field.ty == relative_string_type {
+                        quote! {
+                            #ident: ::std::ffi::CString::new(value.#ident)?
+                        }
+                    } else {
+                        quote! {
+                            #ident: value.#ident
+                        }
+                    }
+                });
+            let cptr_field_initializers = fields
+                .iter()
+                .map(|field| {
+                    let ident = &field.ident.as_ref().unwrap();
+                    if field.ty == absolute_string_type || field.ty == relative_string_type {
+                        quote! {
+                            let #ident = __value.#ident.as_ptr()
+                        }
+                    } else {
+                        quote! {
+                            let #ident = __value.#ident
+                        }
+                    }
+                });
+            let base_field_initializers = fields
+                .iter()
+                .map(|field| {
+                    let ident = &field.ident.as_ref().unwrap();
+
+                    if field.ty == absolute_string_type || field.ty == relative_string_type {
+                        quote! {
+                            #ident: self.#ident.into_string()?
+                        }
+                    } else {
+                        quote! {
+                            #ident: self.#ident
+                        }
+                    }
+                });
+            let assign_field_initializers = fields
+                .iter()
+                .map(|field| {
+                    field.ident.as_ref().unwrap()
+                });
+            quote! {
+
+                #[allow(non_snake_case)]
+                mod #submodule {
+                    pub mod cptr {
+
+                        #[repr(C)]
+                        pub struct #name {
+                            pub __header: ::c_marshalling::Header<super::cstring::#name>,
```

Comment:
The header is redundant. Instead of storing a pointer to the struct in memory letting it own the strings, just move each `CString` with `into_raw` to make the C struct "own" the pointers. (We could implement drop on the c struct to prevent memory leaks though it could be omitted since it will just be moved to C immediately).